src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/io/AprilTagPhotonCameraIO.java

```diff
@@ -0,0 +1,144 @@
+            tagCornerSumY += tagCorner.y;
+        }
+        return new Point(tagCornerSumX / tagCorners.size(), tagCornerSumY / tagCorners.size());
+    }
+
+    /**
+     * Estimates the camera's rotation relative to the apriltag.
+     *
+     * @param result the camera's pipeline result
+     * @return the estimated rotation
+     */
+    private Rotation3d getBestTargetRelativeRotation(PhotonPipelineResult result) {
+        final List<TargetCorner> tagCorners = result.getBestTarget().getDetectedCorners();
+        final Point tagCenter = getTagCenter(tagCorners);
+        if (photonCamera.getCameraMatrix().isPresent())
+            return correctPixelRot(tagCenter, photonCamera.getCameraMatrix().get());
+        return null;
+    }
+
+    /**
+     * Estimates the camera's pose using Solve PNP using as many tags as possible.
+     *
+     * @param result the camera's pipeline result
+     * @return the estimated pose
+     */
+    private Rotation2d getSolvePNPHeading(PhotonPipelineResult result) {
+        if (result.getMultiTagResult().estimatedPose.isPresent) {
+            final Transform3d cameraPoseTransform = result.getMultiTagResult().estimatedPose.best;
+            final Pose3d estimatedPose = new Pose3d().plus(cameraPoseTransform).relativeTo(FieldConstants.APRIL_TAG_FIELD_LAYOUT.getOrigin());
+            return estimatedPose.toPose2d().getRotation();
+        }
+
+        final Transform3d targetToCamera = result.getBestTarget().getBestCameraToTarget().inverse();
+        final Pose3d estimatedPose = FieldConstants.TAG_ID_TO_POSE.get(result.getBestTarget().getFiducialId()).transformBy(targetToCamera);
+        return estimatedPose.toPose2d().getRotation();
+    }
+
+    private int[] getVisibleTagIDs(PhotonPipelineResult result) {
+        final int[] visibleTagIDs = new int[result.getTargets().size()];
+        visibleTagIDs[0] = result.getBestTarget().getFiducialId();
+        int idAddition = 1;
+
+        for (int i = 0; i < visibleTagIDs.length; i++) {
+            final int currentID = result.getTargets().get(i).getFiducialId();
+
+            if (currentID == visibleTagIDs[0]) {
+                idAddition = 0;
+                continue;
+            }
+            visibleTagIDs[i + idAddition] = currentID;
+        }
+
+        return visibleTagIDs;
+    }
+
+    private double getAverageDistanceFromAllTags(PhotonPipelineResult result) {
+        final int tagsSeen = result.getTargets().size();
+        double totalTagDistance = 0;
+
+        for (int i = 0; i < tagsSeen; i++)
+            totalTagDistance += result.getTargets().get(i).getBestCameraToTarget().getTranslation().getNorm();
+
+        return totalTagDistance / tagsSeen;
+    }
+
+    private double getDistanceFromBestTag(PhotonPipelineResult result) {
+        return result.getBestTarget().getBestCameraToTarget().getTranslation().getNorm();
+    }
+
+    private Rotation3d correctPixelRot(Point pixel, Matrix<N3, N3> camIntrinsics) {
```

Comment:
I know this is a copied method, but it's still rlly ugly and I don't like it. Can't you just make it look better?